File: java/p211.java
/* 
 * Solution to Project Euler problem 211
 * Copyright (c) Project Nayuki. All rights reserved.
 * 
 * https://www.nayuki.io/page/project-euler-solutions
 * https://github.com/nayuki/Project-Euler-solutions
 */

import java.util.Arrays;


public final class p211 implements EulerSolution {
	
	public static void main(String[] args) {
		System.out.println(new p211().run());
	}
	
	
	private static final int LIMIT = 64000000;
	
	
	// Requires about 512 MB of memory
	public String run() {
		// sigma2[i] is the sum of the square of i's divisors.
		// For example, sigma2[6] = 1^2 + 2^2 + 3^2 + 6^2 = 20.
		// Computing using a modification of the sieve of Eratosthenes.
		long[] sigma2 = new long[LIMIT];
		Arrays.fill(sigma2, 1, sigma2.length, 1);
		for (int i = 2; i < sigma2.length; i++) {
			if (sigma2[i] == 1) {
				for (int j = i; j < sigma2.length; j += i)
					sigma2[j] *= powerSquareSum(j, i);
			}
		}
		
		long sum = 0;
		SquareTester sqt = new SquareTester(3 * 5 * 7 * 11 * 13 * 17);
		for (int i = 1; i < sigma2.length; i++) {
			if (sqt.isPerfectSquare(sigma2[i]))
				sum += i;
		}
		return Long.toString(sum);
	}
	
	
	// Suppose i is the highest natural number such that k^i divides n.
	// Then this function returns k^(2*0) + k^(2*1) + k^(2*2) + ... + k^(2*i).
	// For example, with n=50 and k=5, i=2 because k^2 = 25 divides n = 50,
	// so the result is k^0 + k^2 + k^4 = 1 + 25 + 625 = 651.
	private static long powerSquareSum(int n, int k) {
		long result = 1;
		long k2 = (long)k * k;
		while (n % k == 0) {
			n /= k;
			result = result * k2 + 1;
		}
		return result;
	}
	
	
	
	// Consider the set of all squared natural numbers, i.e. {0, 1, 4, 9, 16, 25, ...}.
	// When this set is viewed modulo some number n, usually not every residue is in the set.
	// For example, all squares modulo 3 is {0, 1} - so a perfect square modulo 3 is never 2.
	// By choosing a suitably large modulus, we can quickly exclude many numbers that can't be perfect squares.
	private static final class SquareTester {
		
		// isResidue[i] is true iff there exists a natural number k such that i = k^2 mod modulus.
		// Hence for any i, if isResidue[i mod modulus] is false, then i is not a perfect square.
		private boolean[] isResidue;
		
		
		// Any product of unique small prime numbers excluding 2 makes a good modulus
		// that leads to fast tests. But the behavior is correct for any modulus >= 1.
		public SquareTester(int modulus) {
			if (modulus < 1)
				throw new IllegalArgumentException();
			isResidue = new boolean[modulus];
			for (int i = 0; i < modulus; i++)
				isResidue[(int)((long)i * i % modulus)] = true;
		}
		
		
		public boolean isPerfectSquare(long x) {
			// Reject many but not all numbers that aren't a perfect square.
			// This speed optimization can be omitted without affecting correctness.
			if (!isResidue[(int)(x % isResidue.length)])
				return false;
			
			// A complete algorithm for detecting squares
			long y = 0;
			for (long i = 1L << 31; i != 0; i >>>= 1) {
				y |= i;
				if (y > 3037000499L || y * y > x)
					y ^= i;
			}
			return y * y == x;
		}
		
	}
	
}

File: python/p211.py
# 
# Solution to Project Euler problem 211
# Copyright (c) Project Nayuki. All rights reserved.
# 
# https://www.nayuki.io/page/project-euler-solutions
# https://github.com/nayuki/Project-Euler-solutions
# 

import array, eulerlib


def compute():
	LIMIT = 64000000
	
	# Can be any number >= 1, but it's most beneficial to use a product of unique small primes excluding 2
	RESIDUE_TEST = 3 * 5 * 7 * 11 * 13
	
	isresidue = [False] * RESIDUE_TEST
	for i in range(RESIDUE_TEST):
		isresidue[i * i % RESIDUE_TEST] = True
	
	def is_perfect_square(x):
		# Optional optimization: Check if x is a quadratic residue modulo some number.
		# The modulus was chosen to be a product of k primes; in this case, k = 5.
		# If x is a square, then it must be a quadratic residue modulo each prime.
		# For each prime p, there is an approximately half chance that an arbitrary number
		# is a residue mod p. Thus with 5 primes, only about 1/32 of candidates remain.
		# Note that the prime 2 tells us nothing about whether x is a square, so we exclude it.
		return isresidue[x % RESIDUE_TEST] and eulerlib.is_square(x)
	
	
	# Requires at least 640 MB of memory
	sigma2 = list_sigma2(LIMIT - 1)
	ans = sum(i for i in range(1, LIMIT) if is_perfect_square(sigma2[i]))
	return str(ans)


def list_sigma2(n):
	# If i has a prime factor p <= sqrt, then quasiprimefactor[i] = p.
	# Otherwise i > sqrt must be prime, and quasiprimefactor[i] = 0 because i may overflow an int16.
	sqrt = eulerlib.sqrt(n)
	quasiprimefactor = array.array("H", (0 for _ in range(n + 1)))
	
	# Richer version of the sieve of Eratosthenes
	for i in range(2, sqrt + 1):
		if quasiprimefactor[i] == 0:
			quasiprimefactor[i] = i
			for j in range(i * i, n + 1, i):
				if quasiprimefactor[j] == 0:
					quasiprimefactor[j] = i
	
	sigma2 = array.array("Q", (0 for _ in range(n + 1)))
	sigma2[1] = 1
	for i in range(2, len(sigma2)):
		p = quasiprimefactor[i]
		if p == 0:
			p = i
		sum = 1
		j = i
		p2 = p * p
		k = p2
		while j % p == 0:
			sum += k
			j //= p
			k *= p2
		sigma2[i] = sum * sigma2[j]
	return sigma2


if __name__ == "__main__":
	print(compute())
